BEIMA.Backend/DeviceFunctions/UpdateDevice.cs

```diff
@@ -40,42 +42,98 @@ public static class UpdateDevice
-
-                device.SetLocation(
-                    buildingId,
-                    data.Location.Notes,
-                    data.Location.Latitude,
-                    data.Location.Longitude
-                );
-
-                device.SetFields(data.Fields);
-
-                device.SetLastModified(DateTime.UtcNow, "Anonymous");
-                // TODO: include device type attributes
+                reqForm = await req.ReadFormAsync();
+                data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(reqForm["data"]);                              
+                
+                var deviceDocument = mongo.GetDevice(new ObjectId(id));
+                if (deviceDocument == null)
+                {
+                    return new NotFoundObjectResult(Resources.DeviceNotFoundMessage);
+                }
+
+                device = BsonSerializer.Deserialize<Device>(deviceDocument);                
             }
             catch (Exception)
             {
                 return new BadRequestObjectResult(Resources.CouldNotParseBody);
             }
 
+            if (!ObjectId.TryParse(data.DeviceTypeId, out _))
+            {
+                return new BadRequestObjectResult(Resources.InvalidIdMessage);
+            }            
+
+            // Set Data to new values
+            device.DeviceTypeId = ObjectId.Parse(data.DeviceTypeId);
+            device.DeviceTag = data.DeviceTag;
+            device.Manufacturer = data.Manufacturer;
+            device.ModelNum = data.ModelNum;
+            device.SerialNum = data.SerialNum;
+            device.YearManufactured = data.YearManufactured;
+            device.Notes = data.Notes;
+
+            var reqBuildingId = data.Location.BuildingId;
+            ObjectId? buildingId = reqBuildingId != null ? ObjectId.Parse(reqBuildingId) : null;
+            device.SetLocation(
+                buildingId,
+                data.Location.Notes,
+                data.Location.Latitude,
+                data.Location.Longitude
+            );
+
+            device.SetFields(data.Fields);
+
+            var _storage = StorageDefinition.StorageInstance;
+
+            // Check if there is a new photo
+            var updatePhoto = reqForm.Files.Any(file => file.Name == "photos");
+            if (updatePhoto && device.Photo != null)
+            {
+                var result = await _storage.DeleteFile(device.Photo.FileUid);
+                if (!result)
+                {
+                    log.LogInformation($"Failed to delete a file: {device.Photo.FileUid.ToString()}.");
+                }
+                device.Photo = null;
+            }
+
+            // Remove files from device's file list
+            device.Files = device.Files.FindAll(file => !data.DeletedFiles.Contains(file.FileUid));
+
+            // Removed deleted files from storage
+            foreach (var fileUid in data.DeletedFiles)
```

Comment:
Should we consider waiting until after the database has been successfully updated to delete these?

Reply:
That's a really good idea. That way we can ensure the references are removing before deleting the files.